### PR TITLE
Improve the OIDC/OAuth2 documentation

### LINKED ISSUE #10363: Improve Authentication Documentation
The Authentication documentation is spread over several separated guides:
https://quarkus.io/guides/security-jwt
https://quarkus.io/guides/security-openid-connect
https://quarkus.io/guides/security-openid-connect-web-authentication
https://quarkus.io/guides/security-openid-connect-multitenancy
https://quarkus.io/guides/security-keycloak-authorization
 https://quarkus.io/guides/security-oauth2

It's hard to follow, to see the differences, and a lot is still missing in the documentations,
especially about (integration) testing a secured endpoint.

### LINKED ISSUE #10587: Document how to get the roles with quarkus-oidc
Will start working on it immediately after I merge the userinfo PR which is good to go

### PR DESCRIPTION
**Pull request by sberyozkin**
Fixes #10363
Fixes #10587

### CODE REVIEW

**Comment by sberyozkin** [comment]
I've tried to interlink all the OIDC/OAuth2 docs, clarify a few things, also added a section about the roles, etc
Hi Loic - did a tiny update to the OAuth2 guide, I was keen to mention it can verify JWT via the introspection too :-) but I suppose it is better to continue stressing `OAuth2` extension can only verify the opaque tokens to minimize the confusion

File: docs/src/main/asciidoc/security-openid-connect-web-authentication.adoc [hunk 1]
```diff
@@ -3,16 +3,18 @@ This guide is maintained in the main Quarkus repository
...
 https://github.com/quarkusio/quarkus/tree/master/docs/src/main/asciidoc
 ////
-= Quarkus - Using OpenID Connect to Protect Web Applications
+= Quarkus - Using OpenID Connect to Protect Web Applications using Authorization Code Flow.
 
 include::./attributes.adoc[]
+:toc:
 
-This guide demonstrates how to use the OpenID Connect Extension to protect your application using Quarkus, where authentication and authorization are based on tokens issued by OpenId Connect and OAuth 2.0 compliant Authorization Servers such as https://www.keycloak.org/about.html[Keycloak].
+This guide demonstrates how to use Quarkus OpenID Connect Extension to protect your JAX-RS applications using OpenId Connect Authorization Code Flow supported by OpenId Connect compliant Authorization Servers such as https://www.keycloak.org/about.html[Keycloak].
```
Comment (loicmathieu) [inline docs/src/main/asciidoc/security-openid-connect-web-authentication.adoc:11]: Same here, it's not only JAX-RS that you can protect

File: docs/src/main/asciidoc/security-openid-connect-web-authentication.adoc [hunk 2]
```diff
@@ -266,6 +268,21 @@ Note that `AccessTokenCredential` will have to be used if the Access Token issue
...
 Injection of the `JsonWebToken` and `AccessTokenCredential` is supported in both `@RequestScoped` and `@ApplicationScoped` contexts.
 
+== User Info
+
+Set `quarkus.oidc.user-info-required=true` if a UserInfo JSON object from the OIDC userinfo endpoint has to be requested.
+This will make an `io.quarkus.oidc.UserInfo` (which is a simple `javax.json.JsonObject` wrapper) object accessible as a SecurityIdentity `userinfo` attribute.
+
+== Token Claims And SecurityIdentity Roles
+
+The way the roles are mapped to the SecurityIdentity roles from the verified tokens is identical to how it is done for the link:security-openid-connect#token-claims-and-securityidentity-roles[bearer tokens] with the only difference being is that ID JWT token is used as a source of the roles by default.
```
Comment (loicmathieu) [inline docs/src/main/asciidoc/security-openid-connect-web-authentication.adoc:278]: It's not clear for my what `ID JWT token` is ?

**Review by loicmathieu — COMMENTED** [review]
Looks good by I add some questions  / feedbacks

**Comment by loicmathieu** [comment]
@sberyozkin 

> I was keen to mention it can verify JWT via the introspection too :-) but I suppose it is better to continue stressing OAuth2 extension can only verify the opaque tokens to minimize the confusion

I agree, let's keep the oauth2 extension for opaque token only, even if it can of validate a JWT token, we should keep the documentation as simple as possible and document the best way to do things.

File: docs/src/main/asciidoc/security-keycloak-authorization.adoc [hunk 3]
```diff
@@ -3,15 +3,15 @@ This guide is maintained in the main Quarkus repository
 and pull requests should be submitted there:
 https://github.com/quarkusio/quarkus/tree/master/docs/src/main/asciidoc
 ////
-= Quarkus - Using OpenID Connect and Keycloak to Centralize Authorizations
```
Comment (pedroigor) [inline docs/src/main/asciidoc/security-keycloak-authorization.adoc:6]: I would leave as it stands today because this is related to `Centralized Authorization`.

Comment (sberyozkin) [reply docs/src/main/asciidoc/security-keycloak-authorization.adoc:6]: @pedroigor OK, I wanted to make it crystal clear that it is only related (at the moment) to the bearer tokens, but sure, I can drop this title update, since I'm clarifying it further in the document again...

File: docs/src/main/asciidoc/security-jwt.adoc [hunk 4]
```diff
@@ -9,9 +9,9 @@ include::./attributes.adoc[]
 :extension-name: SmallRye JWT
 :mp-jwt: MicroProfile JWT RBAC
 
-This guide explains how your Quarkus application can utilize MicroProfile Json Web Token (link:https://jwt.io/[JWT]) 
-Role-Based Access Control (link:https://en.wikipedia.org/wiki/Role-based_access_control[RBAC]) to provide
-secured access to the JAX-RS endpoints.
+This guide explains how your Quarkus application can utilize MicroProfile JWT (MP JWT) to verify https://tools.ietf.org/html/rfc7519[JSON Web Token]s, represent them as MP JWT `JsonWebToken` and provide secured access to the JAX-RS endpoints using Bearer Token Authorization and https://en.wikipedia.org/wiki/Role-based_access_control[Role-Based Access Control].
```
[the remark this answers is not in the document]
Comment (sberyozkin) [reply docs/src/main/asciidoc/security-jwt.adoc:12]: @loicmathieu thanks for the comments, sure I'll just say `Quarkus endpoints`

File: docs/src/main/asciidoc/security-openid-connect-web-authentication.adoc (hunk 2 above)
Comment (sberyozkin) [reply docs/src/main/asciidoc/security-openid-connect-web-authentication.adoc:278]: @loicmathieu I'll change it to just `ID token`, it is returned, alongside access token, as part of the OIDC code flow, https://openid.net/specs/openid-connect-core-1_0.html#TokenResponse

**Review by loicmathieu — APPROVED** [review]
LGTM

**Review by pedroigor — APPROVED** [review]

**Comment by sberyozkin** [comment]
@loicmathieu @pedroigor Thanks, I'll be following up with more updates around the security and OIDC docs, cheers

**Comment by famod** [comment]
@sberyozkin This seems to have broken the documentation module in `master`?

```
2020-07-13T16:36:52.8471210Z [ERROR] Failed to execute goal org.asciidoctor:asciidoctor-maven-plugin:2.0.0:process-asciidoc (output-pdf) on project quarkus-documentation: Found 92 issue(s) of severity WARN or higher during conversion -> [Help 1]
2020-07-13T16:36:52.8486147Z org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.asciidoctor:asciidoctor-maven-plugin:2.0.0:process-asciidoc (output-pdf) on project quarkus-documentation: Found 92 issue(s) of severity WARN or higher during conversion
2020-07-13T16:36:52.8492076Z     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:215)
2020-07-13T16:36:52.8497310Z     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
2020-07-13T16:36:52.8502280Z     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
2020-07-13T16:36:52.8508954Z     at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
2020-07-13T16:36:52.8515171Z     at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
2020-07-13T16:36:52.8522291Z     at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
2020-07-13T16:36:52.8527841Z     at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
2020-07-13T16:36:52.8531742Z     at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
2020-07-13T16:36:52.8535933Z     at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
2020-07-13T16:36:52.8539799Z     at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
2020-07-13T16:36:52.8543568Z     at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
2020-07-13T16:36:52.8547441Z     at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
2020-07-13T16:36:52.8551124Z     at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
2020-07-13T16:36:52.8555627Z     at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
2020-07-13T16:36:52.8561053Z     at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
2020-07-13T16:36:52.8567013Z     at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
2020-07-13T16:36:52.8570240Z     at java.lang.reflect.Method.invoke (Method.java:566)
2020-07-13T16:36:52.8575456Z     at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
2020-07-13T16:36:52.8580239Z     at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
2020-07-13T16:36:52.8585715Z     at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
2020-07-13T16:36:52.8590279Z     at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
2020-07-13T16:36:52.8597360Z Caused by: org.apache.maven.plugin.MojoExecutionException: Found 92 issue(s) of severity WARN or higher during conversion
2020-07-13T16:36:52.8602471Z     at org.asciidoctor.maven.AsciidoctorMojo.execute (AsciidoctorMojo.java:247)
2020-07-13T16:36:52.8608593Z     at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
2020-07-13T16:36:52.8613592Z     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
2020-07-13T16:36:52.8618580Z     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
2020-07-13T16:36:52.8623583Z     at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
2020-07-13T16:36:52.8629921Z     at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
2020-07-13T16:36:52.8636324Z     at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
2020-07-13T16:36:52.8643464Z     at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
2020-07-13T16:36:52.8648826Z     at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
2020-07-13T16:36:52.8652874Z     at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
2020-07-13T16:36:52.8656941Z     at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
2020-07-13T16:36:52.8660883Z     at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
2020-07-13T16:36:52.8664625Z     at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
2020-07-13T16:36:52.8668303Z     at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
2020-07-13T16:36:52.8671977Z     at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
2020-07-13T16:36:52.8676622Z     at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
2020-07-13T16:36:52.8682252Z     at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
2020-07-13T16:36:52.8688217Z     at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
2020-07-13T16:36:52.8691283Z     at java.lang.reflect.Method.invoke (Method.java:566)
2020-07-13T16:36:52.8696994Z     at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
2020-07-13T16:36:52.8702190Z     at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
2020-07-13T16:36:52.8707367Z     at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
2020-07-13T16:36:52.8712272Z     at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
```

I can reproduce that locally and once I do `git checkout d7fa26d4cd7beababd8f7b93125dd02ac754dbd2` the problem vanishes.

**Comment by gastaldi** [comment]
I can confirm that by reverting commit dfdcad9 the build works again

**Comment by geoand** [comment]
Yup, saw this in one of my PRs as well
